README.md
[![Build Status](https://travis-ci.org/Yelp/hacheck.png)](https://travis-ci.org/Yelp/hacheck)

**hacheck** is a healthcheck-proxying service. It listens on port 3333, speaks HTTP, and has the following API:

    GET /<protocol>/<service_name>/<port>/<query>

This will check the following locations for service state:

 * `/var/spool/hacheck/all`
 * `/var/spool/hacheck/<service_name>`
 * Depending on the value of `<protocol>`:
  * if `http`: `http://localhost:<port>/<query>`
  * if `tcp`: will attempt to connect to port `<port>` on localhost. `<query>` is currently ignored
  * if `spool`: will only check the spool state
  * if `mysql` and the `mysql_username` and `mysql_password` are set, will do a login and quit on the requested mysql port; `<query>` is ignored and no logical database is selected.

When it does query the actual service check endpoint, **hacheck** MAY cache the value of that query for some amount of time

**hacheck** also comes with the command-line utilities `haup`, `hadown`, and `hastatus`. These take a service name and manipulate the spool files, allowing you to pre-emptively mark a service as "up" or "down".

### Yelp/hacheck versus uber/hacheck

This software was originally developed by [Uber](https://github.com/uber/hacheck).
This version has several features not present in the upstream version:

* By setting the `allow_remote_spool_changes` flag in the config file, it becomes possible to POST changes to spool
  status remotely.
  This is useful if there is some software, like [Paasta](github.com/Yelp/paasta) making decisions centrally about when
  to kill tasks.
* If the client sends the header `X-Haproxy-Server-State`, then we use the port specified in that header and ignore the
  port in the URL.
  This is useful for configurations where an HAProxy has a backend with many servers running on different ports, such
  as when your tasks are running under [Marathon](github.com/mesosphere/marathon).
* Spool files(downtimes) can optionally have expirations, and store information about when they were created.
* Spool files(downtimes) can apply to a service on a specific port - useful if you have multiple copies of a service on
  the same box.
* It is possible to specify (via `http_headers_to_copy` in the config file) a set of headers that should be forwarded
  to the service being checked.

### Dependencies

**hacheck** is written in Python and makes extensive use of the [tornado](http://www.tornadoweb.org/en/stable/) asynchronous web framework (specifically, it uses the coroutine stuff in Tornado 3). Unit tests use nose and mock.

It runs on Python 2.6 and above, as well as Python 3.3 and above.

### Use cases

Imagine you want to take down the server `web01` for maintenance. Just SSH to it, then (as root) run `hadown all` and wait however long your HAproxy healthchecking interval is. Do your maintenance, then run `haup all` to put it back in service. So easy!

### Configuration

`hacheck` accepts a `-c` flag which should point to a YAML-formatted configuration file. Some notable properties of this file:
* `cache_time`: The duration for which check results may be cached
* `service_name_header`: If set, the name of a header which will be populated with the service name on HTTP checks
* `log_path`: Either the string `"stdout"`, the string `"stderr"`, or a fully-qualified path to a file to write logs to. Uses a [WatchedFileHandler](http://docs.python.org/2/library/logging.handlers.html#watchedfilehandler) and ought to play nicely with logrotate
* `mysql_username`: username to use when logging into mysql for checks
* `mysql_password`: password to use when logging into mysql for checks
* `rlimit_nofile`: set the NOFILE rlimit. If the string "max", will set the rlimit to the hard rlimit; otherwise, will be interpreted as an integer and set to that value.
* `allow_remote_spool_changes`: whether to allow remote control of spool files.

### Monitoring

`hacheck` exports some useful monitoring stuff at the `/status` endpoint. It also exports a count of requests by source-IP and service name on the `/status/count` endpoint.

If the [mutornadomon](https://github.com/uber/mutornadomon) package is available, `hacheck` will import and use it, exposing standard stats about tornado to localhost at `/mutornadomon`

### License

This work is licensed under the [MIT License](http://opensource.org/licenses/MIT), the contents of which can be found at [LICENSE.txt](LICENSE.txt).


debian/changelog
hacheck (0.10.0-yelp1) lucid; urgency=low

  * Add https support

 -- Matthew Bentley <bentley@yelp.com>  Mon, 18 Jul 2016 15:08:59 -0700

hacheck (0.9.0-yelp6) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Keshav Varma <keshav@yelp.com>  Mon, 11 Jul 2016 16:03:33 -0700

hacheck (0.9.0-yelp5) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Keshav Varma <keshav@yelp.com>  Mon, 11 Jul 2016 15:45:11 -0700

hacheck (0.9.0-yelp4) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Keshav Varma <keshav@yelp.com>  Mon, 11 Jul 2016 15:17:04 -0700

hacheck (0.9.0-yelp3) lucid; urgency=low

  * No longer follow redirects, instead return
  * Fix tornado version to hopefully avoid memory leaks

 -- Tomas Doran <tdoran@yelp.com>  Tue, 09 Feb 2016 01:37:27 -0800

hacheck (0.9.0-yelp2) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Evan Krall <krall@yelp.com>  Tue, 21 Jul 2015 14:48:16 -0800

hacheck (0.9.0-yelp1) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Evan Krall <krall@yelp.com>  Thu, 16 Jul 2015 17:13:33 -0800

hacheck (0.6.1-yelp2) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- John Billings <billings@yelp.com>  Wed, 21 Jan 2015 13:47:21 -0800

hacheck (0.6.1-yelp1) lucid; urgency=low

  * Switch to yelp-hacheck.

 -- John Billings <billings@yelp.com>  Tue, 20 Jan 2015 13:44:39 -0800

hacheck (0.6.1) lucid; urgency=low

  * Initial release.

 -- John Billings <billings@yelp.com>  Wed, 07 Jan 2015 16:12:14 -0800


tests/test_checker.py
import mock
import socket

try:
    from unittest2 import TestCase
except ImportError:
    from unittest import TestCase

import tornado.concurrent
import tornado.web
import tornado.testing

from hacheck import checker
from hacheck import config
from hacheck import spool

se = mock.sentinel


def bind_synchronous_unused_port():
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind(('127.0.0.1', 0))
    port = s.getsockname()[1]
    return s, port


class ReturnTwoHundred(tornado.web.RequestHandler):
    def get(self):
        self.write(b'TEST OK')


class ExpectServiceNameHeader(tornado.web.RequestHandler):
    def get(self):
        self.write(self.request.headers['SName'])


class ReturnFiveOhOne(tornado.web.RequestHandler):
    def get(self):
        self.set_status(501)
        self.write(b'NOPE')


class ReturnThreeOhOne(tornado.web.RequestHandler):
    def get(self):
        self.set_status(301)
        self.set_header('Location', '/')


class EchoParamFoo(tornado.web.RequestHandler):
    def get(self):
        self.write(self.get_argument('foo'))


class TestChecker(TestCase):
    def test_spool_success(self):
        with mock.patch.object(spool, 'is_up', return_value=(True, {})) as is_up_patch:
            fut = checker.check_spool(se.name, se.port, se.query, None, query_params=None, headers={})
            self.assertIsInstance(fut, tornado.concurrent.Future)
            self.assertTrue(fut.done())
            res = fut.result()
            self.assertEqual(res[0], 200)
            is_up_patch.assert_called_once_with(se.name, port=se.port)

    def test_spool_failure(self):
        with mock.patch.object(spool, 'is_up', return_value=(False, {'service': se.service})) as is_up_patch:
            fut = checker.check_spool(se.name, se.port, se.query, None, query_params=None, headers={})
            self.assertEqual(fut.result()[0], 503)
            is_up_patch.assert_called_once_with(se.name, port=se.port)


class BaseTestHTTPHTTPSChecker(object):
    def check_http_https(self, *args, **kwargs):
        raise NotImplementedError

    def get_app(self):
        return tornado.web.Application([
            ('/', ReturnTwoHundred),
            ('/sname', ExpectServiceNameHeader),
            ('/redirect', ReturnThreeOhOne),
            ('/bip', ReturnFiveOhOne),
            ('/echo_foo', EchoParamFoo),
        ])

    @tornado.testing.gen_test
    def test_check_success(self):
        response = yield self.check_http_https("foo", self.get_http_port(), "/", io_loop=self.io_loop, query_params="",
                                               headers={})
        self.assertEqual((200, b'TEST OK'), response)

    @tornado.testing.gen_test
    def test_check_failure(self):
        code, response = yield self.check_http_https("foo", self.get_http_port(), "/bar", io_loop=self.io_loop,
                                                     query_params="", headers={})
        self.assertEqual(404, code)

    @tornado.testing.gen_test
    def test_check_redirect(self):
        code, response = yield self.check_http_https("foo", self.get_http_port(), "/redirect", io_loop=self.io_loop,
                                                     query_params="", headers={})
        self.assertEqual(301, code)

    @tornado.testing.gen_test
    def test_check_failure_with_code(self):
        code, response = yield self.check_http_https("foo", self.get_http_port(), "/bip", io_loop=self.io_loop,
                                                     query_params="", headers={})
        self.assertEqual(501, code)

    @tornado.testing.gen_test
    def test_check_wrong_port(self):
        code, response = yield self.check_http_https("foo", self.get_http_port() + 1, "/", io_loop=self.io_loop,
                                                     query_params="", headers={})
        self.assertEqual(599, code)

    @tornado.testing.gen_test
    def test_service_name_header(self):
        with mock.patch.dict(config.config, {'service_name_header': 'SName'}):
            code, response = yield self.check_http_https('service_name', self.get_http_port(), "/sname",
                                                         io_loop=self.io_loop, query_params="", headers={})
            self.assertEqual(b'service_name', response)

    @tornado.testing.gen_test
    def test_query_params_passed(self):
        response = yield self.check_http_https("foo", self.get_http_port(), "/echo_foo", io_loop=self.io_loop,
                                               query_params="foo=bar", headers={})
        self.assertEqual((200, b'bar'), response)

    @tornado.testing.gen_test
    def test_query_params_not_passed(self):
        response = yield self.check_http_https("foo", self.get_http_port(), "/echo_foo", io_loop=self.io_loop,
                                               query_params="", headers={})
        self.assertEqual(400, response[0])


class TestHTTPChecker(BaseTestHTTPHTTPSChecker, tornado.testing.AsyncHTTPTestCase):
    def check_http_https(self, *args, **kwargs):
        return checker.check_http(*args, **kwargs)


class TestHTTPSChecker(BaseTestHTTPHTTPSChecker, tornado.testing.AsyncHTTPSTestCase):
    def check_http_https(self, *args, **kwargs):
        return checker.check_https(*args, **kwargs)


class TestServer(tornado.tcpserver.TCPServer):
    @tornado.gen.coroutine
    def handle_stream(self, stream, address):
        yield stream.write(b'hello')
        stream.close()


class TestTCPChecker(tornado.testing.AsyncTestCase):
    def setUp(self):
        super(TestTCPChecker, self).setUp()
        socket, port = tornado.testing.bind_unused_port()
        self.server = TestServer(io_loop=self.io_loop)
        self.server.add_socket(socket)
        self.socket = socket
        self.port = port
        unlistened_socket, unlistened_port = bind_synchronous_unused_port()
        self.unlistened_socket = unlistened_socket
        self.unlistened_port = unlistened_port

    def tearDown(self):
        super(TestTCPChecker, self).tearDown()
        try:
            self.server.stop()
            self.socket.close()
        except Exception:
            pass

    @tornado.testing.gen_test
    def test_check_success(self):
        response = yield checker.check_tcp("foo", self.port, None, io_loop=self.io_loop, query_params="", headers={})
        self.assertEqual(200, response[0])

    @tornado.testing.gen_test
    def test_check_failure(self):
        with mock.patch.object(checker, 'TIMEOUT', 1):
            response = yield checker.check_tcp("foo", self.unlistened_port, None, io_loop=self.io_loop, query_params="",
                                               headers={})
            self.assertEqual(response[0], 503)


class TestSMTPServer(tornado.tcpserver.TCPServer):
    def __init__(self, *args, **kwargs):
        super(TestSMTPServer, self).__init__(*args, **kwargs)
        self.never_respond = False

    @tornado.gen.coroutine
    def handle_stream(self, stream, address):
        if self.never_respond:
            stream.close()
            return
        yield stream.write(b'220 some.host MTA NAME\r\n')
        assert b'QUIT\r\n' == (yield stream.read_until(b'\r\n'))
        yield stream.write(b'221 2.0.0 Bye\r\n')
        stream.close()


class TestSMTPChecker(tornado.testing.AsyncTestCase):
    def setUp(self):
        super(TestSMTPChecker, self).setUp()
        socket, port = tornado.testing.bind_unused_port()
        self.server = TestSMTPServer(io_loop=self.io_loop)
        self.server.add_socket(socket)
        self.socket = socket
        self.port = port

    def tearDown(self):
        super(TestSMTPChecker, self).tearDown()
        try:
            self.server.stop()
            self.socket.close()
        except Exception:
            pass

    @tornado.testing.gen_test
    def test_check_success(self):
        response = yield checker.check_smtp("foo", self.port, None, io_loop=self.io_loop, query_params="", headers={})
        self.assertEqual(200, response[0])

    @tornado.testing.gen_test
    def test_check_failure(self):
        with mock.patch.object(self.server, 'never_respond', True):
            response = yield checker.check_smtp("foo", self.port, None, io_loop=self.io_loop, query_params="",
                                                headers={})
            self.assertEqual((503, 'Peer unexpectedly closed connection'), response)


tests/test_application.py
import json
import tempfile
import shutil

from hacheck.compat import nested

import mock
import tornado.concurrent
import tornado.testing
import yaml

from hacheck import main
from hacheck import spool
from hacheck import cache
from hacheck import config
from hacheck import handlers


class ApplicationTestCase(tornado.testing.AsyncHTTPTestCase):
    def setUp(self):
        # flush the cache before every test
        cache.configure()
        self.config_file = tempfile.NamedTemporaryFile(delete=True)
        self.spool = tempfile.mkdtemp()
        mock_config = {
            'cache_time': 100.0,
            'spool_root': self.spool
        }
        self.config_file.write(yaml.dump(mock_config).encode('utf-8'))
        self.config_file.flush()
        spool.configure(spool_root=self.spool)
        handlers._reset_stats()
        super(ApplicationTestCase, self).setUp()

    def tearDown(self):
        if self.config_file:
            self.config_file.close()
        if self.spool:
            shutil.rmtree(self.spool)

    def get_app(self):
        return main.get_app()

    def test_status(self):
        response = self.fetch('/status')
        self.assertEqual('application/json; charset=UTF-8', response.headers['Content-Type'])
        result = json.loads(response.body.decode('utf-8'))
        self.assertGreater(result['uptime'], 0.0)

    def test_status_count(self):
        response = self.fetch('/status/count')
        self.assertEqual('application/json; charset=UTF-8', response.headers['Content-Type'])
        result = json.loads(response.body.decode('utf-8'))
        self.assertEqual(result['service_access_counts'], {})
        with mock.patch.object(spool, 'is_up', return_value=(True, {"reason": b'YES'})):
            response = self.fetch('/spool/foo/1/status')
            self.assertEqual(response.code, 200)
            self.assertEqual(response.body, b'YES')
        response = self.fetch('/status/count')
        result = json.loads(response.body.decode('utf-8'))
        self.assertEqual(result['service_access_counts'], {'foo': {'127.0.0.1': 1}})

    def test_routing(self):
        with mock.patch.object(handlers.HTTPServiceHandler, 'get') as m:
            self.fetch('/http/foo/1/status')
            m.assert_called_once_with('foo', '1', 'status')
        with mock.patch.object(handlers.TCPServiceHandler, 'get') as m:
            self.fetch('/tcp/bar/2')
            m.assert_called_once_with('bar', '2', '')

    def test_spool_checker(self):
        with mock.patch.object(spool, 'is_up', return_value=(True, {"reason": b'YES'})):
            response = self.fetch('/spool/foo/1/status')
            self.assertEqual(response.code, 200)
            self.assertEqual(response.body, b'YES')
        with mock.patch.object(spool, 'is_up', return_value=(False, {"service": "any", "reason": ""})):
            response = self.fetch('/spool/foo/1/status')
            self.assertEqual(response.code, 503)
            self.assertEqual(response.body, b'Service any in down state')
        with mock.patch.object(spool, 'is_up', return_value=(False, {"service": "any", "reason": "just because"})):
            response = self.fetch('/spool/foo/1/status')
            self.assertEqual(response.code, 503)
            self.assertEqual(response.body, b'Service any in down state: just because')
        with mock.patch.object(
            spool,
            'is_up',
            return_value=(False, {"service": "any", "reason": "reason", "expiration": 5, "creation": 4})
        ):
            response = self.fetch('/spool/foo/1/status')
            self.assertEqual(response.code, 503)
            self.assertRegexpMatches(response.body, b'^Service any in down state since 4\.0+ until 5\.0+: reason$')

    def test_calls_all_checkers(self):
        rv1 = tornado.concurrent.Future()
        rv1.set_result((200, b'OK1'))
        rv2 = tornado.concurrent.Future()
        rv2.set_result((200, b'OK2'))
        checker1 = mock.Mock(return_value=rv1)
        checker2 = mock.Mock(return_value=rv2)
        with mock.patch.object(handlers.SpoolServiceHandler, 'CHECKERS', [checker1, checker2]):
            response = self.fetch('/spool/foo/1/status')
            self.assertEqual(200, response.code)
            self.assertEqual(b'OK2', response.body)
            checker1.assert_called_once_with('foo', 1, 'status', io_loop=mock.ANY, query_params='', headers=mock.ANY)
            checker2.assert_called_once_with('foo', 1, 'status', io_loop=mock.ANY, query_params='', headers=mock.ANY)

    def test_passes_headers(self):
        rv1 = tornado.concurrent.Future()
        rv1.set_result((200, b'OK1'))
        checker1 = mock.Mock(return_value=rv1)
        with mock.patch.object(handlers.SpoolServiceHandler, 'CHECKERS', [checker1]):
            response = self.fetch('/spool/foo/1/status')
            self.assertEqual(200, response.code)
            checker1.assert_called_with(
                'foo', 1, 'status', io_loop=mock.ANY, query_params='',
                headers={
                    'Connection': 'close',
                    'Host': mock.ANY,
                    'Accept-Encoding': 'gzip'
                },
            )

    def test_any_failure_fails_all_first(self):
        rv1 = tornado.concurrent.Future()
        rv1.set_result((404, b'NOK1'))
        rv2 = tornado.concurrent.Future()
        rv2.set_result((200, b'OK2'))
        checker1 = mock.Mock(return_value=rv1)
        checker2 = mock.Mock(return_value=rv2)
        with mock.patch.object(handlers.SpoolServiceHandler, 'CHECKERS', [checker1, checker2]):
            response = self.fetch('/spool/foo/2/status')
            self.assertEqual(404, response.code)
            self.assertEqual(b'NOK1', response.body)

    def test_any_failure_fails_all_second(self):
        rv1 = tornado.concurrent.Future()
        rv1.set_result((200, b'OK1'))
        rv2 = tornado.concurrent.Future()
        rv2.set_result((404, b'NOK2'))
        checker1 = mock.Mock(return_value=rv1)
        checker2 = mock.Mock(return_value=rv2)
        with mock.patch.object(handlers.SpoolServiceHandler, 'CHECKERS', [checker1, checker2]):
            response = self.fetch('/spool/foo/2/status')
            self.assertEqual(404, response.code)
            self.assertEqual(b'NOK2', response.body)

    def test_weird_code(self):
        # test that unusual HTTP codes are rewritten to 503s
        rv = tornado.concurrent.Future()
        rv.set_result((6000, 'this code is weird'))
        checker = mock.Mock(return_value=rv)
        with mock.patch.object(handlers.HTTPServiceHandler, 'CHECKERS', [checker]):
            response = self.fetch('/http/uncached-weird-code/80/status')
            self.assertEqual(503, response.code)

    def test_haproxy_server_state(self):
        rv = tornado.concurrent.Future()
        rv.set_result((200, b'OK'))
        checker = mock.Mock(return_value=rv)
        server_state = 'UP 2/3; addr=srv1; port=1234; name=bck/srv2; node=lb1; weight=1/2; scur=13/22; qcur=0'
        with mock.patch.object(handlers.HTTPServiceHandler, 'CHECKERS', [checker]):
            response = self.fetch('/http/foo/1/status', headers={'X-Haproxy-Server-State': server_state})
            self.assertEqual(200, response.code)
            args, _ = checker.call_args
            assert args[1] == 1234

    def test_old_haproxy_server_state_ignored(self):
        rv = tornado.concurrent.Future()
        rv.set_result((200, b'OK'))
        checker = mock.Mock(return_value=rv)
        server_state = 'UP 2/3; name=bck/srv2; node=lb1; weight=1/2; scur=13/22; qcur=0'
        with mock.patch.object(handlers.HTTPServiceHandler, 'CHECKERS', [checker]):
            response = self.fetch('/http/foo/1/status', headers={'X-Haproxy-Server-State': server_state})
            self.assertEqual(200, response.code)
            args, _ = checker.call_args
            assert args[1] == 1

    def test_option_parsing(self):
        with nested(
            mock.patch('sys.argv', ['ignorethis', '-c', self.config_file.name, '--spool-root', 'foo']),
            mock.patch.object(tornado.ioloop.IOLoop, 'instance'),
            mock.patch.object(cache, 'configure'),
            mock.patch.object(main, 'get_app'),
            mock.patch.object(spool, 'configure')) \
                as (_1, _2, cache_configure, _3, spool_configure):
            main.main()
            spool_configure.assert_called_once_with(spool_root='foo')
            cache_configure.assert_called_once_with(cache_time=100)

    def test_show_recent(self):
        handlers.seen_services.clear()
        response = self.fetch('/spool/foo/1/status')
        self.assertEqual(200, response.code)
        response = self.fetch('/recent')
        b = json.loads(response.body.decode('utf-8'))
        self.assertEqual(
            b,
            {
                'seen_services': [['foo', {'code': 200, 'ts': mock.ANY, 'remote_ip': '127.0.0.1'}]],
                'threshold_seconds': 600
            })
        response = self.fetch('/recent?threshold=20')
        b = json.loads(response.body.decode('utf-8'))
        self.assertEqual(
            b,
            {
                'seen_services': [['foo', {'code': 200, 'ts': mock.ANY, 'remote_ip': '127.0.0.1'}]],
                'threshold_seconds': 20
            })

    def test_remote_spool_check_forbidden(self):
        with mock.patch.dict(config.config, {'allow_remote_spool_changes': False}):
            response = self.fetch('/spool/foo/1/status', method='POST', body="")
            self.assertEqual(response.code, 403)

    def test_spool_post(self):
        with nested(
            mock.patch.dict(config.config, {'allow_remote_spool_changes': True}),
            mock.patch.object(spool, 'up'),
            mock.patch.object(spool, 'down'),
                ) as (_1, spool_up, spool_down):

            response = self.fetch('/spool/foo/0/status', method='POST', body="status=up")
            self.assertEqual(response.code, 200)
            spool_up.assert_called_once_with('foo', port=None)

            response = self.fetch('/spool/foo/1234/status', method='POST', body="status=down&reason=because")
            self.assertEqual(response.code, 200)
            spool_down.assert_called_once_with('foo', reason='because', port=1234, expiration=None, creation=None)

            spool_down.reset_mock()
            response = self.fetch('/spool/foo/1234/status', method='POST',
                                  body="status=down&reason=because&expiration=1&creation=2")
            self.assertEqual(response.code, 200)
            spool_down.assert_called_once_with('foo', reason='because', port=1234, expiration=1, creation=2)


hacheck/main.py
import logging
import optparse
import signal
import time
import types
import sys
import resource

import tornado.ioloop
import tornado.httpserver
import tornado.web
from tornado.log import access_log

from . import cache
from . import config
from . import handlers
from . import spool

try:
    from mutornadomon.config import initialize_mutornadomon
except ImportError:
    initialize_mutornadomon = None


class DummyTimeout(object):
    callback = None
DUMMY = DummyTimeout()


def log_request(handler):
    # log requests at INFO instead of WARNING for all status codes
    request_time = 1000.0 * handler.request.request_time()
    access_log.debug("%d %s %.2fms", handler.get_status(),
                     handler._request_summary(), request_time)


def get_app():
    return tornado.web.Application([
        (r'/http/([.a-zA-Z0-9_-]+)/([0-9]+)/(.*)', handlers.HTTPServiceHandler),
        (r'/https/([.a-zA-Z0-9_-]+)/([0-9]+)/(.*)', handlers.HTTPSServiceHandler),
        (r'/tcp/([.a-zA-Z0-9_-]+)/([0-9]+)/?(.*)', handlers.TCPServiceHandler),
        (r'/mysql/([.a-zA-Z0-9_-]+)/([0-9]+)/?(.*)', handlers.MySQLServiceHandler),
        (r'/smtp/([.a-zA-Z0-9_-]+)/([0-9]+)/?(.*)', handlers.SMTPServiceHandler),
        (r'/spool/([.a-zA-Z0-9_-]+)/([0-9]+)/?(.*)', handlers.SpoolServiceHandler),
        (r'/recent', handlers.ListRecentHandler),
        (r'/status/count', handlers.ServiceCountHandler),
        (r'/status', handlers.StatusHandler),
    ], start_time=time.time(), log_function=log_request)


def remove_timeout(self, timeout):
    timeout.callback = None
    timeout = DUMMY
    self._cancellations += 1


def setrlimit_nofile(soft_target):
    current_soft, current_hard = resource.getrlimit(resource.RLIMIT_NOFILE)
    if soft_target == 'max':
        desired_fd_limit = (current_hard, current_hard)
    elif soft_target > current_hard:
        raise ValueError('Targeted NOFILE rlimit %d is greater than hard limit %d' % (soft_target, current_hard))
    else:
        desired_fd_limit = (soft_target, current_hard)
    resource.setrlimit(resource.RLIMIT_NOFILE, desired_fd_limit)


def main():
    parser = optparse.OptionParser()
    parser.add_option(
        '-c',
        '--config-file',
        default=None,
        help='Path to a YAML config file'
    )
    parser.add_option(
        '-p',
        '--port',
        default=[],
        type=int,
        action='append',
        help='Port to listen on. May be repeated. If not passed, defaults to :3333.'
    )
    parser.add_option(
        '-B',
        '--bind-address',
        default='0.0.0.0',
        help='Address to listen on. Defaults to %default'
    )
    parser.add_option(
        '--spool-root',
        default='/var/spool/hacheck',
        help='Root for spool for service states (default %default)'
    )
    parser.add_option(
        '-v',
        '--verbose',
        default=False,
        action='store_true'
    )
    opts, args = parser.parse_args()
    if opts.config_file is not None:
        config.load_from(opts.config_file)

    if not opts.port:
        opts.port = [3333]
    if config.config['rlimit_nofile'] is not None:
        setrlimit_nofile(config.config['rlimit_nofile'])

    # set up logging
    log_path = config.config['log_path']
    level = logging.DEBUG if opts.verbose else logging.WARNING
    if log_path == 'stdout':
        handler = logging.StreamHandler(sys.stdout)
    elif log_path == 'stderr':
        handler = logging.StreamHandler(sys.stderr)
    else:
        handler = logging.handlers.WatchedFileHandler(log_path)
    fmt = logging.Formatter(logging.BASIC_FORMAT, None)
    handler.setFormatter(fmt)
    logging.getLogger().addHandler(handler)
    logging.getLogger().setLevel(level)

    # application stuff
    cache.configure(cache_time=config.config['cache_time'])
    spool.configure(spool_root=opts.spool_root)
    application = get_app()
    ioloop = tornado.ioloop.IOLoop.instance()
    ioloop.remove_timeout = types.MethodType(remove_timeout, ioloop)
    server = tornado.httpserver.HTTPServer(application, io_loop=ioloop)

    if initialize_mutornadomon is not None:
        mutornadomon_collector = initialize_mutornadomon(application, io_loop=ioloop)
    else:
        mutornadomon_collector = None

    def stop(*args):
        if mutornadomon_collector is not None:
            mutornadomon_collector.stop()
        ioloop.stop()

    for port in opts.port:
        server.listen(port, opts.bind_address)
    for sig in (signal.SIGTERM, signal.SIGQUIT, signal.SIGINT):
        signal.signal(sig, stop)
    ioloop.start()
    return 0


if __name__ == '__main__':
    sys.exit(main())


hacheck/handlers.py
import collections
import logging
import time

import tornado.ioloop
import tornado.httputil
import tornado.httpclient
import tornado.gen
import tornado.web

from . import cache
from . import checker
from . import config
from . import spool

log = logging.getLogger('hacheck')

StatusResponse = collections.namedtuple('StatusResponse', ['code', 'remote_ip', 'ts'])

if hasattr(collections, 'Counter'):
    Counter = collections.Counter  # fast
else:
    def Counter():
        return collections.defaultdict(lambda: 0)

seen_services = {}
service_count = collections.defaultdict(Counter)
last_statuses = {}


def _reset_stats():
    seen_services.clear()
    service_count.clear()
    last_statuses.clear()


class StatusHandler(tornado.web.RequestHandler):
    def get(self):
        stats = {}
        stats['cache'] = cache.get_stats()
        stats['uptime'] = time.time() - self.settings['start_time']
        self.set_status(200)
        self.write(stats)


class ListRecentHandler(tornado.web.RequestHandler):
    def get(self):
        now = time.time()
        recency_threshold = int(self.get_argument('threshold', 10 * 60))
        response = []
        for service_name, t in seen_services.items():
            if now - t > recency_threshold:
                continue
            last_status = last_statuses.get(service_name, None)
            if last_status is not None:
                last_status = last_status._asdict()
            response.append((service_name, last_status))
        self.write({
            'seen_services': list(sorted(response)),
            'threshold_seconds': recency_threshold
        })


class ServiceCountHandler(tornado.web.RequestHandler):
    def get(self):
        self.write({'service_access_counts': dict(service_count)})


class BaseServiceHandler(tornado.web.RequestHandler):
    CHECKERS = []

    def maybe_get_port_from_haproxy_server_state(self):
        """
        Look for the 'X-Haproxy-Server-State' header and try to parse out the
        'port' value.

        Note that only very recent versions of HAProxy support sending
        the port in the send-state header. In particular you need
        commit 514061c414080701cb046171041a2d00859660e8 from haproxy dev.

        Example server state header:

            X-Haproxy-Server-State: UP 2/3; address=srv2; port=1234;
              name=bck/srv2; node=lb1; weight=1/2; scur=13/22; qcur=0

        returns: the string-typed port if found, else None.
        """

        server_state = self.request.headers.get('X-Haproxy-Server-State', '')

        parts = server_state.split(';')
        parts = [part.strip() for part in parts]
        parts = [part.split('=') for part in parts]
        parts = [part for part in parts if len(part) == 2]

        return dict(parts).get('port')

    @tornado.web.asynchronous
    @tornado.gen.coroutine
    def get(self, service_name, port, query):
        port = self.maybe_get_port_from_haproxy_server_state() or port

        seen_services[service_name] = time.time()
        service_count[service_name][self.request.remote_ip] += 1
        with cache.maybe_bust(self.request.headers.get('Pragma', '') == 'no-cache'):
            port = int(port)
            last_message = ""
            querystr = self.request.query
            for this_checker in self.CHECKERS:
                code, message = yield this_checker(
                    service_name,
                    port,
                    query,
                    io_loop=tornado.ioloop.IOLoop.current(),
                    query_params=querystr,
                    headers=self.request.headers,
                )
                last_message = message
                if code > 200:
                    last_statuses[service_name] = StatusResponse(code, self.request.remote_ip, time.time())
                    if code in tornado.httputil.responses:
                        self.set_status(code)
                    else:
                        self.set_status(503)
                    self.write(message)
                    self.finish()
                    break
            else:
                last_statuses[service_name] = StatusResponse(200, self.request.remote_ip, time.time())
                self.set_status(200)
                self.write(last_message)
                self.finish()


class SpoolServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool]

    def post(self, service_name, port, query):
        if not config.config['allow_remote_spool_changes']:
            self.set_status(403)
            self.write('remote spool changes are not enabled')
            return

        port = int(port) or None
        status = self.get_argument('status')

        if status == 'up':
            spool.up(service_name, port=port)
        elif status == 'down':
            expiration = self.get_argument('expiration', None)
            if expiration is not None:
                expiration = float(expiration)
            reason = self.get_argument('reason')
            creation = self.get_argument('creation', None)
            if creation is not None:
                creation = float(creation)
            spool.down(service_name, reason=reason, port=port, expiration=expiration, creation=creation)
        else:
            self.set_status(400)
            self.write("status must be up or down")
            return

        self.set_status(200)
        self.write("")


class HTTPServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_http]


class HTTPSServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_https]


class TCPServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_tcp]


class MySQLServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_mysql]


class SMTPServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_smtp]
